##### **Understanding recursion** #####

##### Definition #####
- Function calling itself
- Almost all the situation where we use loops
  - substitute the loops using recursion
- Can solve problems that seem very complex at first

##### Example - factorial #####

In [1]:
def factorial(n):
    result = 1
    while n > 1:
        result = n * result
        n -= 1
    return result

In [2]:
factorial(5)

120

##### Example - identifying the base case #####
- Add a condition
  - ensures that our algorithm does't execute forever
- **Factorial base case** -> n = 1

In [4]:
# Avoid infinite recursion, the first thing is to identify the base case
def factorial_recursion(n):
    if n == 1:
        return 1
    else:
        return n * factorial_recursion(n-1) # recursive call

In [5]:
print(factorial_recursion(5))

120


In [7]:
%timeit factorial(5)

105 ns ± 0.204 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [8]:
%timeit factorial_recursion(5)

166 ns ± 0.518 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


##### How recursion works #####
- Computer uses a **stack** to keep track of the funtions
  - **Call stack**
- `factorial(5)` starts

- Before `factorial(5)` finishes ->\
  `factorial(4)` starts, \
  the computer needs to know that `factorial(5)` didn't finish, so it pushes the information into the call stack.

- Before `factorial(4)` finishes ->\
  `factorial(3)` starts

- Before `factorial(3)` finishes ->\
  `factorial(2)` starts

- Before `factorial(2)` finishes ->\
  `factorial(1)` starts

- `factorial(1)` finishes
  - returns 1

- `factorial(2)` finishes
  - returns 2

- `factorial(3)` finishes
  - returns 6

- `factorial(4)` finishes
  - returns 24

- `factorial(5)` finishes
  - returns 120
  


##### Dynamic programming #####
- Optimization technique
- Mainly applied to recursion
- Can reduce the complexity of recursive algorithms
- Used for:
  - Any problem that can be divided into smaller subproblems
  - Subproblems overlap
- Solutions of subproblems are saved, avoiding the need to recalculate
  - Memorization

In [9]:
def fibonacci(n):
  # Define the base case
  if n <= 1:
    return n
  else:
    # Call recursively to fibonacci
    return fibonacci(n-1) + fibonacci(n-2)
    
print(fibonacci(6))

8


In [10]:
# saving the solutions of the subproblems in the cache variable to avoid recalculating.

cache = [None]*(100)

def fibonacci(n): 
    if n <= 1:
        return n
    
    # Check if the value exists
    if not cache[n]:
        # Save the result in cache
        cache[n] = fibonacci(n-1) + fibonacci(n-2)
    
    return cache[n]
    

print(fibonacci(6))

8


In [11]:
print(cache[:10])

[None, None, 1, 2, 3, 5, 8, None, None, None]
